File: 백준/Platinum/1168. 요세푸스 문제 2/요세푸스 문제 2.cc
#include <stdio.h>

int N, K;
int tree[262144];
int start, leafSize;

void update(int index, int diff)
{
    index += start;

    while (index > 1)
    {
        tree[index] += diff;
        index /= 2;
    }
}

int getNumber(int index)
{
    int treeIdx = 1;
    while (treeIdx < leafSize)
    {
        if (index <= tree[treeIdx * 2])
            treeIdx *= 2;
        else
        {
            index -= tree[treeIdx * 2];
            treeIdx = treeIdx * 2 + 1;
        }
    }

    return treeIdx - start;
}

int main()
{
    scanf("%d %d", &N, &K);

    for (leafSize = 1; leafSize <= N; leafSize <<= 1)
        ;
    start = leafSize - 1;

    for (int i = 1; i <= N; i++)
        update(i, 1);

    int index, remain;

    index = K, remain = N;

    printf("<");
    for (int i = 1; i <= N - 1; i++)
    {
        int number = getNumber(index);

        printf("%d, ", number);

        update(number, -1);

        index = index - 1 + K;
        remain = remain - 1;
        index = (index - 1) % remain + 1;
    }

    printf("%d>\n", getNumber(1));

    return 0;
}

File: 백준/Platinum/1168. 요세푸스 문제 2/요세푸스 문제 2.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.StringTokenizer;

public class Main {
    static int N, K;
    static int[] tree = new int[200005];

    static void update(int idx, int val) {
        while (idx <= 200004) {
            tree[idx] += val;
            idx += (idx & -idx);
        }
    }

    static int sum(int idx) {
        int ret = 0;
        while (idx > 0) {
            ret += tree[idx];
            idx -= (idx & -idx);
        }
        return ret;
    }

    static int getNumber(int index) {
        int left = 1, right = N;
        while (left <= right) {
            int mid = (left + right) / 2;
            int count = sum(mid);
            if (count >= index)
                right = mid - 1;
            else
                left = mid + 1;
        }
        return left;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        K = Integer.parseInt(st.nextToken());

        for (int i = 1; i <= N; i++)
            update(i, 1);

        int index = K;

        StringBuilder sb = new StringBuilder();
        sb.append("<");
        for (int i = 1; i <= N - 1; i++) {
            int number = getNumber(index);

            sb.append(number).append(", ");

            update(number, -1);

            index = index - 1 + K;
            index = (index - 1) % (N - i) + 1;
        }

        sb.append(getNumber(1)).append(">");

        System.out.println(sb.toString());
    }
}
